File: day1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func abs(a, b int) int{
    tmp := a - b
    if tmp > 0{
        return tmp
    }
    return -1*tmp
}

func main() {
	file, err := os.Open("./input.txt")
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	lhs := make([]int, 1000)
	rhs := make([]int, 1000)
    rval_counter := make(map[int]int)

	idx := 0
	for scanner.Scan() {
		line := scanner.Text()
		values := strings.Split(line, "   ")
		lval, lerr := strconv.Atoi(values[0])
		check(lerr)
		lhs[idx] = lval
		rval, rerr := strconv.Atoi(values[1])
        rval_counter[rval] += 1
		check(rerr)
		rhs[idx] = rval
		idx += 1
	}

	sort.Ints(lhs)
	sort.Ints(rhs)
	differnce := 0
	for i, val := range lhs {
		differnce += abs(val, rhs[i])
	}
    similarity := 0
    for _, val := range lhs {
        similarity += val * rval_counter[val]
    }
	fmt.Println("Difference : ", differnce)
    fmt.Println("Similarity : ", similarity)
}

File: day1/main.py
#!/usr/bin/env python3
### Day 1 AoC

from collections import Counter
import os.path

INPUT = os.path.join(os.path.abspath(os.path.dirname(__file__)), "input.txt")

def main() -> None:
    
    lhs = []
    rhs = []
    with open(INPUT, 'r') as f:
        for line in f.readlines():
            data = line.strip().split()
            lhs.append(int(data[0]))
            rhs.append(int(data[1]))

    lhs.sort()
    rhs.sort()
    count = Counter(rhs)
    distances = [abs(lh - rhs[i]) for i, lh in enumerate(lhs)]
    print(f"Total difference is {sum(distances)}")
    similarity = [val * count[val] for val in lhs]
    print(f"Similarity score is {sum(similarity)}")

if __name__ == "__main__":
    main()
